File: python/lib/rover.py
from compass import Compass

class Rover(object):
  def __init__(self, color, location_str, command_str):
    self.color   = color
    x, y, heading = location_str.split(' ')
    self.x_coord = int(x)
    self.y_coord = int(y)
    self.compass = Compass(heading)
    self.commands = command_str

    pass
  
  def complete_mission(self):
    for c in self.commands:
      self.execute_command(c)
  
    pass
  
  def execute_command(self, command):
    if command == "L":
      self.turn_left()
    elif command == "R":
      self.turn_right()
    elif command == "M":
      self.go_forward()
    else:
      raise ValueError("incorrect command! - {0}".format(c))
  
  def current_position(self):
    return "x: {0}, y: {1}, heading: {2}".format(self.x_coord, self.y_coord, self.compass.direction())

  def go_forward(self):
    heading = self.compass.direction()
 
    if heading == Compass.NORTH:
      self.y_coord += 1
    elif heading == Compass.EAST:
      self.x_coord += 1
    elif heading == Compass.SOUTH:
      self.y_coord -= 1
    elif heading == Compass.WEST:
      self.x_coord -= 1
    else:
      raise ValueError("incorrect heading! - {0}".format(heading))
    pass

  def turn_left(self):
    self.compass.rotate_left()
    pass

  def turn_right(self):
    self.compass.rotate_right()
    pass

File: elixir/lib/rover_challenge/rover.ex
defmodule RoverChallenge.Rover do
  defstruct [:x_coord, :y_coord, :compass]
  alias RoverChallenge.Compass

  @moduledoc """
  Documentation for RoverChallenge.Rover
  """

  def init(location_str) when is_bitstring(location_str) do
    [x_coord, y_coord, heading] = location_str |> String.split(" ")
    init(x_coord, y_coord, heading)
  end

  def init(x, y, heading) do
    %__MODULE__{
      x_coord: String.to_integer(x),
      y_coord: String.to_integer(y),
      compass: Compass.init(heading)
    }
  end

  def execute_command(rover, command) do
    case command do
      "L" -> turn_left(rover)
      "R" -> turn_right(rover)
      "M" -> go_forward(rover)
      _   -> {:error, "the command '#{command}' is invalid"}
    end
  end

  def coordinates(rover) do
    {rover.x_coord, rover.y_coord, Compass.direction(rover.compass)}
  end

  defp turn_right(rover), do: {:ok, %{rover | compass: Compass.rotate_right(rover.compass)}}
  defp turn_left(rover),  do: {:ok, %{rover | compass: Compass.rotate_left(rover.compass)}}
  defp go_forward(rover) do
    case Compass.direction(rover.compass) do
      "N" -> {:ok, %{rover | y_coord: rover.y_coord + 1}}
      "E" -> {:ok, %{rover | x_coord: rover.x_coord + 1}}
      "S" -> {:ok, %{rover | y_coord: rover.y_coord - 1}}
      "W" -> {:ok, %{rover | x_coord: rover.x_coord - 1}}
      _   -> {:error, "your compass is broken - '#{rover.compass.directions}'"}
    end
  end
end
